### PR TITLE
Rule Based Classifier

### PR DESCRIPTION
**Pull request by DomingoLenar**

### CODE REVIEW

**Review by Benny-Gil — DISMISSED** [review]
kindly double check your work as all metrics are perfect with score of 100, it may be too overfitted or the implementation maybe wrong.


image

**Review by VladTemp27 — CHANGES_REQUESTED** [review]
Please redo your rule-based classification. For more feedback please read my comments thank you.

**Review by Benny-Gil — COMMENTED** [review]
looks better

File: rule_based.ipynb [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+    "import numpy as np\n",
+    "import pandas as pd\n",
+    "import matplotlib.pyplot as plt\n",
+    "import seaborn as sns\n",
+    "from sklearn.model_selection import train_test_split\n",
+    "from sklearn.metrics import accuracy_score, classification_report, confusion_matrix\n",
+    "from collections import Counter\n",
+    "\n",
+    "# Fix the file path to use proper escape characters\n",
+    "csv_path = r\"D:\\student-classification-model\\graded_exams.csv\"  # Using raw string to avoid escape sequence issues\n",
```
Comment (VladTemp27) [inline rule_based.ipynb:18]: Do not use absolute path

File: rule_based.ipynb [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+    "    # Group by the selected features and analyze grade distribution\n",
+    "    grouped = data.groupby(features_to_analyze)\n",
+    "    \n",
+    "    for name, group in grouped:\n",
+    "        # Only consider combinations with enough samples\n",
+    "        if len(group) >= min_samples:\n",
+    "            # Count occurrences of each grade\n",
+    "            grade_counts = Counter(group['grade'])\n",
+    "            \n",
+    "            # Find the most common grade\n",
```
Comment (VladTemp27) [inline rule_based.ipynb:100]: Document this more, add justification

File: rule_based.ipynb [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class AdvancedRuleClassifier:\n",
+    "    def __init__(self, subject):\n",
+    "        self.subject = subject\n",
+    "        self.rules = []\n",
+    "        self.default_grade = None\n",
+    "        \n",
+    "    def fit(self, X_train, y_train):\n",
```
Comment (VladTemp27) [inline rule_based.ipynb:125]: I think `build` would be a more appropriate function name

File: rule_based.ipynb [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+    "def plot_scatter(y_true, y_pred, subject):\n",
+    "    plt.figure(figsize=(8, 6))\n",
+    "    plt.scatter(y_true, y_pred, alpha=0.5, edgecolors='k')\n",
+    "    plt.plot([min(y_true), max(y_true)], [min(y_true), max(y_true)], linestyle=\"--\", color=\"red\", label=\"Perfect Prediction\")\n",
+    "    \n",
+    "    plt.xlabel(\"Actual Grades\")\n",
+    "    plt.ylabel(\"Predicted Grades\")\n",
+    "    plt.title(f\"Scatter Plot of Actual vs. Predicted Grades ({subject})\")\n",
+    "    plt.legend()\n",
+    "    plt.show()\n",
```
Comment (VladTemp27) [inline rule_based.ipynb:354]: I think don't use scatterplot as a medium or visualization for comparison, use it for looking at correlation.

**Review by VladTemp27 — CHANGES_REQUESTED** [review]
Please view my comments. We're getting there

File: rule_based.ipynb (hunk 4 above)
Comment (DomingoLenar) [reply rule_based.ipynb:354]: image
should it look more like this?, though this is a lot of scatter plots as it has a lot of categorical values to correlate it from

Comment (VladTemp27) [reply rule_based.ipynb:354]: Learn how to properly use a scatter plot

**Comment by DomingoLenar** [comment]
Comments are now implemented and pushed to the rule_based branch, please review the updated rule_based branch.

File: rule_based.ipynb [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+    "    To ensure reliability, a minimum sample threshold is applied, preventing rare and \n",
+    "    potentially misleading patterns from affecting decision-making. This improves the \n",
+    "    consistency of the model and reduces noise in the classification process.\n",
+    "\n",
+    "    The function is particularly useful in rule-based systems, as it generates \n",
+    "    human-readable classification rules that are easy to interpret and apply. The \n",
+    "    output is a dictionary that maps feature combinations to their most common grade, \n",
+    "    along with a distribution of grades observed for that combination.\n",
+    "    \"\"\"\n",
+    "    results = {}\n",
```
Comment (VladTemp27) [inline rule_based.ipynb:99]: Looking at the algorithm isnt this the same as apiori, otherwise can you please differentiate them

File: rule_based.ipynb [hunk 6]
```diff
@@ -0,0 +1,456 @@
...
+    "    \n",
+    "    plt.xlabel(\"Actual Grades (Numeric)\")\n",
+    "    plt.ylabel(\"Predicted Grades (Numeric)\")\n",
+    "    plt.title(f\"Correlation between Actual and Predicted Grades ({subject})\\nCorrelation: {correlation:.2f}\")\n",
+    "    plt.show()\n",
+    "\n",
+    "# Example usage\n",
+    "plot_correlation(y_test_math, y_pred_math, \"Math\")\n",
+    "plot_correlation(y_test_reading, y_pred_reading, \"Reading\")\n",
+    "plot_correlation(y_test_writing, y_pred_writing, \"Writing\")\n"
```
Comment (VladTemp27) [inline rule_based.ipynb:378]: I will repeat, do not correlate the output. correlate characteristics to the output. Again please learn how to use correlation

**Review by VladTemp27 — CHANGES_REQUESTED** [review]

File: rule_based.ipynb (hunk 6 above)
Comment (DomingoLenar) [reply rule_based.ipynb:378]: I will be removing scatter plot for rule based classifier as I am correlating categorical values, as scatter plot is used to correlate two numerical values, and not two categorical values

File: rule_based.ipynb (hunk 5 above)
Comment (DomingoLenar) [reply rule_based.ipynb:99]: I will be adding an explanation to the code block.

Comment (Benny-Gil) [reply rule_based.ipynb]: based on understanding, [Classification Based on Association Rules](https://ceur-ws.org/Vol-2204/paper6.pdf), which is a type of a rule-based classifier, often use apriori algorithm to discover the class association rules, however there maybe something wrong with threshhold affecting the accuracy of the model.

**Comment by Benny-Gil** [comment]
@DomingoLenar 

Kindly look into the following rule-based learning algorithms:
[First-Order Inductive Learner (FOIL) Algorithm](https://www.geeksforgeeks.org/first-order-inductive-learner-foil-algorithm/)
[CBA (Classification Based on Association Rules) ](https://ceur-ws.org/Vol-2204/paper6.pdf) -> this seems to be the current implementation
[Ripper Algorithm](https://www.geeksforgeeks.org/ripper-algorithm/)

**Comment by DomingoLenar** [comment]
I've added @Benny-Gil's suggestions on following the algorithms, specifically on CBA. Code has been pushed, please check.

**Review by VladTemp27 — APPROVED** [review]
Okay na to
